src/main/java/com/google/api/generator/engine/ast/GeneralForStatement.java

```diff
@@ -61,22 +65,24 @@ public static Builder builder() {
 
   @AutoValue.Builder
   public abstract static class Builder {
-    public abstract Builder setInitializationExpr(Expr initializationExpr);
-
-    public abstract Builder setBody(List<Statement> body);
-
-    // Private.
-    abstract Builder setLocalVariableExpr(VariableExpr variableExpr);
-
-    abstract Builder setMaxSizeExpr(Expr maxSizeExpr);
-
-    abstract VariableExpr localVariableExpr();
+    // Private setter.
+    abstract Builder setInitializationExpr(AssignmentExpr initializationExpr);
+    // Private setter.
+    abstract Builder setTerminationExpr(Expr terminationExpr);
+    // Private setter.
+    abstract Builder setIncrementExpr(Expr incrementExpr);
+    // Private setter.
+    abstract Builder setBody(List<Statement> body);
 
     abstract GeneralForStatement autoBuild();
 
     // Type-checking will be done in the sub-expressions.
     public GeneralForStatement build() {
-      VariableExpr varExpr = localVariableExpr();
+      GeneralForStatement generalForStatement = autoBuild();
+      AssignmentExpr initializationExpr = generalForStatement.initializationExpr();
+      Expr terminationExpr = generalForStatement.terminationExpr();
+      Expr incrementExpr = generalForStatement.incrementExpr();
+      VariableExpr varExpr = initializationExpr.variableExpr();
       Preconditions.checkState(
           varExpr.scope().equals(ScopeNode.LOCAL),
```

Comment:
This should have been a `final` check only, could you please fix this as well?

Reply:
I am not sure if I get the point here. Variable declared in initialization should be a local variable, variable scope private, public, protected, static, are not allowed, but final is allowed in below case with i++ invalid of "cannot assign a value to final variable 'i'"
```
    for (final int i = 0; i < 10; i++) {
      System.out.println("i = " + i);
    } 
```

Reply:
I got the point. I will update the doc with variable type checking.